src/main/java/org/jboss/logging/Log4j2Logger.java

```diff
@@ -52,7 +52,7 @@ protected void doLog(final Level level, final String loggerClassName, final Obje
         if (this.logger.isEnabled(translatedLevel)) {
             try {
                 this.logger.logMessage(loggerClassName, translatedLevel, null,
-                        (parameters == null || parameters.length == 0) ? this.messageFactory.newMessage(message) : this.messageFactory.newMessage(String.valueOf(message), parameters),
+                        (parameters == null || parameters.length == 0) ? this.messageFactory.newMessage(String.valueOf(message)) : this.messageFactory.newMessage(String.valueOf(message), parameters),
```

Comment:
I think we should do something like:
```
this.messageFactory.newMessage((message instanceof String) ? (String) message : message)
```

Reply:
Or actually maybe should use a `CharSequence` instead of a string.

Reply:
I think the gist of this PR is to maintain consistency in the following two cases:
* non-empty parameter list is provided;
* empty parameter list is provided;
We solved the first case by
```
this.messageFactory.newMessage(String.valueOf(message), parameters)
```
because of the following method signature:
```
    public Message newMessage(String message, Object... params) {
        return new MessageFormatMessage(message, params);
    }
```
I think we should strive to invoke the same method in case 2. For that reason, I think the original PR makes sense. Why?
because if this method is invoked, we wanna avoid invoking `this.messageFactory.newMessage(message)` for it would lead to the following statement:
```
public Message newMessage(Object message) {
        return new ObjectMessage(message);
    }
```
However, in log4j2 (not in log4j), there exists the following methods in `org.apache.logging.log4j.message.ObjectMessage` class:
```
    public Object getParameter() {
        return this.obj;
    }

    public Object[] getParameters() {
        return new Object[]{this.obj};
    }
```
That is the root cause of the bug. Suppose the following message is provided without any parameter:
```
binding parameter [someName] as [VARCHAR] - [{}]
```
the message per se would be treated as the parameter for the `{}` placeholder and the final logging statement would be:
```
binding parameter [someName] as [VARCHAR] - [binding parameter [someName] as [VARCHAR] - [{}]]
```
Thus the bug and the gist of this PR.
`String.valueOf()` is much stronger than String casting and I do think its usage should be retained consistently.

Reply:
Another approach is the following statement:
```
this.messageFactory.newMessage(String.valueOf(message), parameters == null ? new String[0] : parameters),
```
Anyway, we should strive to come up with a String out of message. Failing to do that with empty parameter always easily bring about the above example bug.

Reply:
I adopted the above approach. It enjoys the benefit that same method would be used consistently regardless of whether parameter is empty or not.

Reply:
Sorry for all the delays on this. I finally had a chance to look at what the `ObjectMessage` does and I think your initial approach was correct. I apologize for that. If we always use the `MessageFormatMessage` then it's processed through a formatter where as just using the `SimpleMessage` just the message will be returned and won't be processed through a formatter.

Reply:
anyway, you can choose either the initial commit or the latter one by git tweaking. Either approach is fine to me!